---
title: "Create OmicSignature"
output: rmarkdown::html_vignette
author: "Vanessa Mengze Li"
date: "10/24/2024"
vignette: >
  %\VignetteIndexEntry{Create OmicSignature}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r message=FALSE, warning=FALSE}
devtools::load_all(".")
library(dplyr)
```

<font color = "#000000">

## 1. Cheatsheet

An `OmicSignature` object contains three parts:\

-   **metadata**, a list.\
    Required fields:\
    "**signature_name**", "**organism**", "**direction_type**", "**assay_type**", "**phenotype**".\
    Recommended optional fields, if applicable:\
    "platform", "sample_type", "description", "covariates", "score_cutoff", "adj_p_cutoff".\

-   **signature**, a data frame.\
    Required columns:\
    "**probe_id**" (unique identifier for each feature)\
    "**feature_name**" (e.g. ENSEMBL ID, Uniprot ID)\
    Required for bi-directional and categorical signatures: "**direction**"\
    Recommended optional column, if applicable: "score"\

-   **difexp** (optional), a data frame of differential expression analysis results.\
    Required columns:\
    "**probe_id**" (unique identifier for each feature)\
    "**feature_name**" (e.g. ENSEMBL ID, Uniprot ID)\
    "**score**"\
    At least one of the following: "**p_value**", "**q_value**", or "**adj_p**".\
    Recommended optional column, if applicable: "gene_symbol"\

Create the object:\

```         
OmS <- OmicSignature$new(
  metadata = metadata,
  signature = signature,
  difexp = difexp
)
```

## 2. Create an `OmicSignature` Object Step-by-Step

The example provided below is from an experiment evaluating the effects of Myc reduced expression by comparing liver profiles of 24-month old *Myc*$^{+/+}$ *vs.* *Myc*$^{+/-}$ mice. This is a *bi-directional* signature example, since it contains up and down regulated features (genes). For ease of exposition, only the top 1000 genes are here included.\

### 2.1. Metadata

A list with the following required fields:\
"**signature_name**", "**organism**", "**direction_type**", "**assay_type**", "**phenotype**".\
Not required, but highly recommended fields:\
"**platform**", "**sample_type**", "**description**", "**covariates**", "**score_cutoff**", "**adj_p_cutoff**", "**logfc_cutoff**".\

Option 1: Create `metadata` by hand. This is not recommended, because typos can occur.\

```         
metadata <- list(
  "signature_name" = Myc_reduce_mice_liver_24m,
  "organism" = "Mus Musculus",
  "sample_type" = "liver",
  "phenotype" = "Myc_reduce",
  "direction_type" = "bi-directional",
  "assay_type" = "transcriptomics", 
  "platform" = "transcriptomics by array"
)
```

Option 2: Use `createMetadata()` (recommended).\
This function helps remind you of the built-in attributes. The full list of current built-in attributes is shown [here](https://montilab.github.io/OmicSignature/reference/createMetadata.html).\
You can also provide your own customized attributes using the "others" field.\

```{r}
metadata <- createMetadata(
  # required attributes:
  signature_name = "Myc_reduce_mice_liver_24m",
  organism = "Mus Musculus",
  direction_type = "bi-directional",
  assay_type = "transcriptomics",
  phenotype = "Myc_reduce",

  # optional and recommended:
  covariates = "none",
  description = "mice Myc haploinsufficient (Myc(+/-))",
  platform = "transcriptomics by array", 
  sample_type = "liver", # use BRENDA ontology

  # optional cut-off attributes.
  # specifying them can facilitate the extraction of signatures.
  logfc_cutoff = NULL,
  p_value_cutoff = NULL,
  adj_p_cutoff = 0.05,
  score_cutoff = 5,

  # other optional built-in attributes:
  keywords = c("Myc", "KO", "longevity"),
  cutoff_description = NULL,
  author = NULL,
  PMID = 25619689,
  year = 2015,

  # example of customized attributes:
  others = list("animal_strain" = "C57BL/6")
)
```

#### 2.1.1 "phenotype"

"phenotype" is free text, one or two-word description of the experimental conditions, such as a drug treatment, a gene knock out, or a clinical trait. For example, *age* or *breast cancer*.\
Providing a detailed (free text) "description" is highly recommended. For instance, it may include information about how the treatment was administered and how each group was defined.\

#### 2.1.2 "sample_type" and "platform"

While both "sample_type" and "platform" are free text, it is recommended that predefined terms are used. Thus, "sample_type" should be a BRENDA ontology term, and "platform" should be one of a set of predefine platforms, if appropriate. You can search for predefined terms via the `searchSampleType` and `searchPlatform` functions. See ["Sample Type & Platform Info"](https://montilab.github.io/OmicSignature/articles/SampleType.html) for details.\

#### 2.1.3 "direction_type"

`direction_type` must be one of the following:\

-   "uni-directional". Only a list of significant feature names is available. Examples including "genes mutated in a disease" and "markers of a cell type".\

-   "bi-directional". Significant features can be grouped into "up" and "down" categories. For example, when comparing treatment *vs.* control groups, some features will be higher ("up", or "+") and some will be lower ("down" or "-") in the treatment group. When the phenotype is a continuous trait, such as age, some features will increase ("up", or "+") with age, while others will decrease ("down", or "-") with age.\

-   "categorical". Used with multi-valued categorical phenotypes (e.g., "A" *vs.* "B" *vs.* "C"), usually analyzed by ANOVA.\

#### 2.1.4 "assay_type"

`assay_type` is one of the following:\
- "transcriptomics" (e.g. RNA-seq, micro-array)\
- "proteomics"\
- "metabolomics"\
- "methylomics"\
- "genetic_variations" (e.g. SNP, GWAS)\
- "DNA_binding_sites" (e.g. ChIP-seq)\
- "other"\

### 2.2. Differential expression analysis results (difexp)

A differential expression dataframe is **optional** but **highly recommended** if available. It facilitates downstream signature extraction.\

`difexp` is a dataframe with the following **required** columns:\
"**probe_id**", "**feature_name**", "**score**", along with at least one of the following: "**p_value**", "**q_value**", or "**adj_p**".\

"**probe_id**" is a unique identifier, usually a platform-specific identifider (e.g., probe IDs for Affymetrix microarrays, or aptamer IDs for SomaScan assays). If not provided, it will be automatically generated.\
"**feature_name**" is a name that identifies each feature, examples include ENSEMBL IDs for transcripts, UniProt IDs for proteins, and Refmet IDs for metabolites. To better identify the features, it is recommended to add an additional annotation column(s), e.g., "gene_symbol". For metabolite features, it is recommended to include multiple annotation columns if available, e.g., HMDB ID and InChI key.\
"**p_value**", "**q_value**", or "**adj_p**" refers to the p- or q-value representing the significance of each feature.\
"score" is a numeric value that indicates the importance or significance of a feature. Depending on how the signature was derived, this can be the t-statistics, log-fold change, Z-score, or other summary statistics.\

Here we use an example from the differential expression analysis using the `limma` package.

```{r}
# Version reading from a txt file
# difexp <- read.table(
#   file.path(
#     system.file("extdata", package = "OmicSignature"),
#     "difmatrix_Myc_mice_liver_24m.txt"
#   ),
#   header = TRUE, sep = "\t", stringsAsFactors = FALSE
# )
# Version reading from a binary file
difexp <- readRDS(
  file.path(
    system.file("extdata", package = "OmicSignature"),
    "difmatrix_Myc_mice_liver_24m.rds"
  )
)
head(difexp)
```

Manually change the column names to match the requirement. The built-in function `replaceDifexpCol()` designed to replace some frequently used alternative column names.\

```{r}
difexp <- difexp %>% rename(feature_name = ensembl)
colnames(difexp) <- replaceDifexpCol(colnames(difexp))
head(difexp)
```

### 2.3. Signature

**signature** is a dataframe with column **"probe_id"** and **"feature_name"**. If the signature is "bi-directional" or "categorical" (specified in `direction_type` in `metadata`), then column **"direction"** is also required.\
An optional column "score" is recommended when applicable.\

Option 1: Extract signature from difexp.\
In this example, we create a bi-directional signature from the difexp using the `score_cutoff` and `adj_p_cutoff` specified in the metadata.\

```{r}
signature <- difexp %>%
  dplyr::filter(abs(score) > metadata$score_cutoff & adj_p < metadata$adj_p_cutoff) %>%
  dplyr::select(probe_id, feature_name, score) %>%
  dplyr::mutate(direction = ifelse(score > 0, "+", "-"))
head(signature)
```

Function `standardizeSigDF()` can help remove duplicated and empty names.

```{r}
signature <- standardizeSigDF(signature)
head(signature)
```

Option 2: Manually create signature dataframe.\
For uni-directional signature:\

```         
signature <- data.frame("probe_id" = c(1, 2, 3), "feature_name" = c("gene1", "gene2", "gene3"))
```

For bi-directional signature:\

```         
signature <- data.frame(
  "probe_id" = c(1, 2, 3),
  "feature_name" = c("gene1", "gene2", "gene3"),
  "score" = c(0.45, -3.21, 2.44),
  "direction" = c("+", "-", "+")
)
```

For categorical signature:\

```         
signature <- data.frame(
  "probe_id" = c(1, 2, 3, 4),
  "feature_name" = c("gene1", "gene2", "gene3", "gene4"),
  "score" = c(0.45, -3.21, 2.44, -2.45),
  "direction" = c("group1", "group1", "group2", "group3")
)
```

### 2.4. Create the `OmicSignature` object

```{r}
OmS <- OmicSignature$new(
  metadata = metadata,
  signature = signature,
  difexp = difexp
)
```

Set `print_message` = `TRUE` to see the messages.\

```{r}
OmS <- OmicSignature$new(
  metadata = metadata,
  signature = signature,
  difexp = difexp,
  print_message = TRUE
)
```

See the created object information:

```{r}
print(OmS)
```

Use new criteria to extract significant features:\
(this does *not* change the `signature` saved in the object)

```{r}
OmS$extractSignature("abs(score) > 10; adj_p < 0.01")
```

Besides save and read OmicSignature object in `.rds` format, you can export the object as a json text file.\

```         
saveRDS(OmS, "Myc_reduce_mice_liver_24m_OmS.rds")
writeJson(OmS, "Myc_reduce_mice_liver_24m_OmS.json")
```

See more in ["Functionalities of OmicSignature"](https://montilab.github.io/OmicSignature/articles/FunOmS.html) section.\

## 3. Create an `OmicSignature` from `difexp` and `metadata`

You can by-pass the generating signature process once you are an expert. Simply provide cutoffs (e.g. `adj_p_cutoff` and `score_cutoff`) in the `metadata`, make sure `difexp` has "adj_p" and "score" columns, and use `OmicSigFromDifexp()` to automatically extract significant features and create the `OmicSignature` object.\

```{r}
OmS1 <- OmicSigFromDifexp(difexp, metadata)
OmS1
```

## </font>
