<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Create OmicSignature • OmicSignature</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Create OmicSignature">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">OmicSignature</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html">Home</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/ObjectStructure.html">Structure of OmicSignature &amp; OmicSignatureCollection</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>---- OmicSignature ----</h6></li>
    <li><a class="dropdown-item" href="../articles/CreateOmS.html">Create an OmicSignature</a></li>
    <li><a class="dropdown-item" href="../articles/FunOmS.html">Functionalities, imports and exports</a></li>
    <li><a class="dropdown-item" href="../articles/SampleType.html">Sample Type &amp; Platform Info</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>---- OmicSignatureCollection ----</h6></li>
    <li><a class="dropdown-item" href="../articles/CreateOmSC.html">Create an OmicSignatureCollection</a></li>
    <li><a class="dropdown-item" href="../articles/FunOmSC.html">Functionalities of OmicSignatureCollection</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Create OmicSignature</h1>
                        <h4 data-toc-skip class="author">Vanessa Mengze
Li</h4>
            
            <h4 data-toc-skip class="date">2025-10-07</h4>
      

      <div class="d-none name"><code>CreateOmS.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/load_all.html" class="external-link">load_all</a></span><span class="op">(</span><span class="st">"."</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
<p><font color="#000000"></font></p>
<div class="section level2">
<h2 id="cheatsheet">1. Cheatsheet<a class="anchor" aria-label="anchor" href="#cheatsheet"></a>
</h2>
<p>An <code>OmicSignature</code> object contains three parts:<br></p>
<ul>
<li><p><strong>metadata</strong>, a list.<br>
Required fields:<br>
“<strong>signature_name</strong>”, “<strong>organism</strong>”,
“<strong>direction_type</strong>”, “<strong>assay_type</strong>”,
“<strong>phenotype</strong>”, “<strong>author</strong>”.<br>
Recommended optional fields, if applicable:<br>
“platform”, “sample_type”, “description”, “covariates”, “score_cutoff”,
“adj_p_cutoff”.<br></p></li>
<li><p><strong>signature</strong>, a data frame.<br>
Required columns:<br>
“<strong>probe_id</strong>” (unique identifier for each feature)<br>
“<strong>feature_name</strong>” (e.g. ENSEMBL ID, Uniprot ID)<br>
Required for bi-directional and categorical signatures:
“<strong>group_label</strong>”<br>
Recommended optional column, if applicable: “score”<br></p></li>
<li><p><strong>difexp</strong> (optional), a data frame of differential
expression analysis results.<br>
Required columns:<br>
“<strong>probe_id</strong>” (unique identifier for each feature)<br>
“<strong>feature_name</strong>” (e.g. ENSEMBL ID, Uniprot ID)<br>
“<strong>score</strong>”<br>
Required for bi-directional and categorical signatures:
“<strong>group_label</strong>”<br>
At least one of the following: “<strong>p_value</strong>”,
“<strong>q_value</strong>”, or “<strong>adj_p</strong>”.<br>
Recommended optional column, if applicable: “gene_symbol”<br></p></li>
</ul>
<p>Create the object:<br></p>
<pre><code><span><span class="va">OmS</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/OmicSignature.html">OmicSignature</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span></span>
<span>  metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>  signature <span class="op">=</span> <span class="va">signature</span>,</span>
<span>  difexp <span class="op">=</span> <span class="va">difexp</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="create-an-omicsignature-object-step-by-step">2. Create an <code>OmicSignature</code> Object Step-by-Step<a class="anchor" aria-label="anchor" href="#create-an-omicsignature-object-step-by-step"></a>
</h2>
<p>The example provided below is from an experiment evaluating the
effects of Myc reduced expression by comparing liver profiles of
24-month old
<em>Myc</em><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mrow><mi>+</mi><mi>/</mi><mo>+</mo></mrow></msup><annotation encoding="application/x-tex">^{+/+}</annotation></semantics></math><em>vs.</em>
<em>Myc</em><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mrow><mi>+</mi><mi>/</mi><mo>−</mo></mrow></msup><annotation encoding="application/x-tex">^{+/-}</annotation></semantics></math>
mice. This is a <em>bi-directional</em> signature example, since it is a
comparison between two groups, and contains up and down regulated
features (genes). For ease of exposition, only the top 1000 genes are
here included.<br></p>
<div class="section level3">
<h3 id="metadata">2.1. Metadata<a class="anchor" aria-label="anchor" href="#metadata"></a>
</h3>
<p>A list with the following required fields:<br>
“<strong>signature_name</strong>”, “<strong>organism</strong>”,
“<strong>direction_type</strong>”, “<strong>assay_type</strong>”,
“<strong>phenotype</strong>”, “<strong>author</strong>”.<br>
Not required, but highly recommended fields:<br>
“<strong>platform</strong>”, “<strong>sample_type</strong>”,
“<strong>description</strong>”, “<strong>covariates</strong>”,
“<strong>score_cutoff</strong>”, “<strong>adj_p_cutoff</strong>”,
“<strong>logfc_cutoff</strong>”.<br></p>
<p>Option 1: Create <code>metadata</code> by hand. This is not
recommended, because typos can occur.<br></p>
<pre><code><span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="st">"signature_name"</span> <span class="op">=</span> <span class="va">Myc_reduce_mice_liver_24m</span>,</span>
<span>  <span class="st">"organism"</span> <span class="op">=</span> <span class="st">"Mus musculus"</span>,</span>
<span>  <span class="st">"sample_type"</span> <span class="op">=</span> <span class="st">"liver"</span>,</span>
<span>  <span class="st">"phenotype"</span> <span class="op">=</span> <span class="st">"Myc_reduce"</span>,</span>
<span>  <span class="st">"direction_type"</span> <span class="op">=</span> <span class="st">"bi-directional"</span>,</span>
<span>  <span class="st">"assay_type"</span> <span class="op">=</span> <span class="st">"transcriptomics"</span>, </span>
<span>  <span class="st">"platform"</span> <span class="op">=</span> <span class="st">"transcriptomics by array"</span>,</span>
<span>  <span class="st">"author"</span> <span class="op">=</span> <span class="st">"Researcher1"</span></span>
<span><span class="op">)</span></span></code></pre>
<p>Option 2: Use <code><a href="../reference/createMetadata.html">createMetadata()</a></code> (recommended).<br>
This function helps remind you of the built-in attributes. The full list
of current built-in attributes is shown <a href="https://montilab.github.io/OmicSignature/reference/createMetadata.html">here</a>.<br>
You can also provide your own customized attributes using the “others”
field.<br></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createMetadata.html">createMetadata</a></span><span class="op">(</span></span>
<span>  <span class="co"># required attributes:</span></span>
<span>  signature_name <span class="op">=</span> <span class="st">"Myc_reduce_mice_liver_24m"</span>,</span>
<span>  organism <span class="op">=</span> <span class="st">"Mus musculus"</span>,</span>
<span>  direction_type <span class="op">=</span> <span class="st">"bi-directional"</span>,</span>
<span>  assay_type <span class="op">=</span> <span class="st">"transcriptomics"</span>,</span>
<span>  phenotype <span class="op">=</span> <span class="st">"Myc_reduce"</span>,</span>
<span>  author <span class="op">=</span> <span class="st">"Researcher1"</span>,</span>
<span></span>
<span>  <span class="co"># optional and recommended:</span></span>
<span>  covariates <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  description <span class="op">=</span> <span class="st">"mice Myc haploinsufficient (Myc(+/-))"</span>,</span>
<span>  platform <span class="op">=</span> <span class="st">"transcriptomics by array"</span>,</span>
<span>  sample_type <span class="op">=</span> <span class="st">"liver"</span>, <span class="co"># use BRENDA ontology</span></span>
<span></span>
<span>  <span class="co"># optional cut-off attributes.</span></span>
<span>  <span class="co"># specifying them can facilitate the extraction of signatures.</span></span>
<span>  logfc_cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  p_value_cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  adj_p_cutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  score_cutoff <span class="op">=</span> <span class="fl">5</span>,</span>
<span></span>
<span>  <span class="co"># other optional built-in attributes:</span></span>
<span>  keywords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Myc"</span>, <span class="st">"KO"</span>, <span class="st">"longevity"</span><span class="op">)</span>,</span>
<span>  cutoff_description <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  PMID <span class="op">=</span> <span class="fl">25619689</span>,</span>
<span>  year <span class="op">=</span> <span class="fl">2015</span>,</span>
<span></span>
<span>  <span class="co"># example of customized attributes:</span></span>
<span>  others <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"animal_strain"</span> <span class="op">=</span> <span class="st">"C57BL/6"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="phenotype">2.1.1 “phenotype”<a class="anchor" aria-label="anchor" href="#phenotype"></a>
</h4>
<p>“‘phenotype’ is free text, usually a one- or two-word description of
the experimental condition or trait under study, such as a drug
treatment, a gene knockout, or a clinical characteristic. Examples
include <em>age</em>, <em>breast cancer</em>, and <em>drug X</em>.</p>
<p>Providing a detailed (free text) “description” is highly recommended.
For instance, it may include information about how the treatment was
administered and how each group was defined. The character limit of the
description is 65,535 (16-bit storage), space included.<br></p>
</div>
<div class="section level4">
<h4 id="organism">2.1.2 “organism<a class="anchor" aria-label="anchor" href="#organism"></a>
</h4>
<p>“organism” is free text. We provide a pre-defined list of common
organisms, which you can search using <code><a href="../reference/searchOrganism.html">searchOrganism()</a></code>.
Other entries are allowed, but please use standard naming conventions
(e.g., “Homo sapiens”, “Mus musculus”) to ensure consistency.<br></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/searchOrganism.html">searchOrganism</a></span><span class="op">(</span><span class="st">"homo"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Homo sapiens"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="sample_type-and-platform">2.1.3 “sample_type” and “platform”<a class="anchor" aria-label="anchor" href="#sample_type-and-platform"></a>
</h4>
<p>While both “sample_type” and “platform” are free text, it is
recommended that predefined terms are used. Thus, “sample_type” should
be a BRENDA ontology term, and “platform” should be one of a set of
predefined platforms, if appropriate. You can search for predefined
terms via the <code>searchSampleType</code> and
<code>searchPlatform</code> functions. See <a href="https://montilab.github.io/OmicSignature/articles/SampleType.html">“Sample
Type &amp; Platform Info”</a> for details.<br></p>
</div>
<div class="section level4">
<h4 id="direction_type">2.1.4 “direction_type”<a class="anchor" aria-label="anchor" href="#direction_type"></a>
</h4>
<p><code>direction_type</code> must be one of the following:<br></p>
<ul>
<li><p>“uni-directional”. Only a list of significant feature names is
available. Examples including “genes mutated in a disease” and “markers
of a specific cell type”.<br></p></li>
<li><p>“bi-directional”. Significant features are derived from
comparison of two groups, or a single continuous trait. Thus, the
resulting significant features can be grouped into two categories. For
example, when comparing treatment <em>vs.</em> control groups, some
features will be higher and some will be lower in the treatment group.
When the phenotype is a continuous trait, such as age, some features
will increase with age, while others will decrease with age.<br></p></li>
<li><p>“categorical”. Used with multi-valued categorical phenotypes
(e.g., “A” <em>vs.</em> “B” <em>vs.</em> “C”), usually analyzed by
ANOVA.<br></p></li>
</ul>
</div>
<div class="section level4">
<h4 id="assay_type">2.1.5 “assay_type”<a class="anchor" aria-label="anchor" href="#assay_type"></a>
</h4>
<p><code>assay_type</code> is one of the following:<br>
- “transcriptomics” (e.g. RNA-seq, micro-array)<br>
- “proteomics”<br>
- “metabolomics”<br>
- “methylomics”<br>
- “genetic_variations” (e.g. SNP, GWAS)<br>
- “DNA_binding_sites” (e.g. ChIP-seq)<br>
- “other”<br></p>
<p>You can also use <code><a href="../reference/searchAssayType.html">searchAssayType()</a></code> to see the list
above.<br></p>
</div>
</div>
<div class="section level3">
<h3 id="signature">2.2. Signature<a class="anchor" aria-label="anchor" href="#signature"></a>
</h3>
<p><strong>signature</strong> is a dataframe with the columns
<strong>“probe_id”</strong> and <strong>“feature_name”</strong>. If the
signature is bi-directional or categorical (as specified in
<code>direction_type</code> within <code>metadata</code>), an additional
column, <strong>“group_label”</strong>, is also required.<br>
An optional column <strong>“score”</strong> is highly recommended when
applicable.<br><br>
“<strong>probe_id</strong>” is a unique identifier, usually a
platform-specific identifider (e.g., probe IDs for Affymetrix
microarrays, or aptamer IDs for SomaScan assays). If not provided, it
will be automatically generated.<br>
“<strong>feature_name</strong>” is a name that identifies each feature,
examples include ENSEMBL IDs for transcripts, UniProt IDs for proteins,
and Refmet IDs for metabolites. To better identify the features, it is
recommended to add an additional annotation column(s), e.g.,
“gene_symbol”. For metabolite features, it is recommended to include
multiple annotation columns if available, e.g., HMDB ID and InChI
key.<br><strong>“group_label”</strong> is a factor column. This column indicates
the experimental group in which a feature is more highly expressed or
more significant. For example, if the analysis identifies genes
differentially expressed in <em>Treatment v.s. Control</em>, all gene
features with a positive score should be labeled “Treatment” in this
column, indicating their higher expression in the Treatment group.
Features with negative scores should be labeled “Control”. Similarly, if
the analysis concerns protein expression changes associated with BMI,
protein features with positive scores should be labeled “higher BMI”,
while those with negative scores should be labeled “lower BMI”.<br><br></p>
<p><strong>Option 1</strong>: Extract signature from a differential
analysis results table.<br>
In this example, we extract a bi-directional signature from a
<code>difexp</code> object (see next section) using the
<code>score_cutoff</code> and <code>adj_p_cutoff</code> specified in the
metadata. A <code>difexp</code> object is a properly formatted data
frame reporting the results of a differential analysis (e.g., by
Limma).<br></p>
<pre><code><span><span class="co">#&gt;   probe_id  logfc  mean  score p_value adj_p      b       feature_name</span></span>
<span><span class="co">#&gt; 1 10345228 -0.167 7.106 -1.470   0.186 0.560 -5.866 ENSMUSG00000103746</span></span>
<span><span class="co">#&gt; 2 10354534  0.041 4.351  0.520   0.620 0.870 -6.780 ENSMUSG00000060715</span></span>
<span><span class="co">#&gt; 3 10354529 -0.175 4.955 -0.941   0.379 0.731 -6.458 ENSMUSG00000043629</span></span>
<span><span class="co">#&gt; 4 10346337  0.025 8.621  0.188   0.857 0.962 -6.911 ENSMUSG00000038323</span></span>
<span><span class="co">#&gt; 5 10353792 -0.025 6.063 -0.284   0.785 0.936 -6.885 ENSMUSG00000045815</span></span>
<span><span class="co">#&gt; 6 10350848 -0.055 7.595 -0.630   0.549 0.836 -6.712 ENSMUSG00000049881</span></span></code></pre>
<p>We can then extract the significant features as follows:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">signature</span> <span class="op">&lt;-</span> <span class="va">difexp</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">score</span><span class="op">)</span> <span class="op">&gt;</span> <span class="va">metadata</span><span class="op">$</span><span class="va">score_cutoff</span> <span class="op">&amp;</span> <span class="va">adj_p</span> <span class="op">&lt;</span> <span class="va">metadata</span><span class="op">$</span><span class="va">adj_p_cutoff</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">probe_id</span>, <span class="va">feature_name</span>, <span class="va">score</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>group_label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">score</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="st">"MYC Reduce"</span>, <span class="st">"WT"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">signature</span><span class="op">)</span></span>
<span><span class="co">#&gt;   probe_id       feature_name   score group_label</span></span>
<span><span class="co">#&gt; 1 10346882 ENSMUSG00000025964  -6.990          WT</span></span>
<span><span class="co">#&gt; 2 10353878 ENSMUSG00000067653  -7.867          WT</span></span>
<span><span class="co">#&gt; 3 10349648 ENSMUSG00000004552  14.762  MYC Reduce</span></span>
<span><span class="co">#&gt; 4 10355278 ENSMUSG00000062209   6.083  MYC Reduce</span></span>
<span><span class="co">#&gt; 5 10353192 ENSMUSG00000025932  10.487  MYC Reduce</span></span>
<span><span class="co">#&gt; 6 10345762 ENSMUSG00000026072 -13.543          WT</span></span></code></pre></div>
<p>Function <code><a href="../reference/standardizeSigDF.html">standardizeSigDF()</a></code> can help remove duplicated
and empty names.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">signature</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/standardizeSigDF.html">standardizeSigDF</a></span><span class="op">(</span><span class="va">signature</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">signature</span><span class="op">)</span></span>
<span><span class="co">#&gt;   probe_id       feature_name   score group_label</span></span>
<span><span class="co">#&gt; 1 10349648 ENSMUSG00000004552  14.762  MYC Reduce</span></span>
<span><span class="co">#&gt; 2 10345762 ENSMUSG00000026072 -13.543          WT</span></span>
<span><span class="co">#&gt; 3 10353192 ENSMUSG00000025932  10.487  MYC Reduce</span></span>
<span><span class="co">#&gt; 4 10355259 ENSMUSG00000061816 -10.315          WT</span></span>
<span><span class="co">#&gt; 5 10351477 ENSMUSG00000102418   8.818  MYC Reduce</span></span>
<span><span class="co">#&gt; 6 10353878 ENSMUSG00000067653  -7.867          WT</span></span></code></pre></div>
<p><strong>Option 2</strong>: Manually create signature dataframe.<br>
For uni-directional signature:<br></p>
<pre><code><span><span class="va">signature</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="st">"probe_id"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>, <span class="st">"feature_name"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene1"</span>, <span class="st">"gene2"</span>, <span class="st">"gene3"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>For bi-directional signature:<br></p>
<pre><code><span><span class="va">signature</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  <span class="st">"probe_id"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  <span class="st">"feature_name"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene1"</span>, <span class="st">"gene2"</span>, <span class="st">"gene3"</span><span class="op">)</span>,</span>
<span>  <span class="st">"score"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.45</span>, <span class="op">-</span><span class="fl">3.21</span>, <span class="fl">2.44</span><span class="op">)</span>,</span>
<span>  <span class="st">"group_label"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Treatment"</span>, <span class="st">"Control"</span>, <span class="st">"Treatment"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
<p>For multi-categorical signature:<br></p>
<pre><code><span><span class="va">signature</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  <span class="st">"probe_id"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">4</span><span class="op">)</span>,</span>
<span>  <span class="st">"feature_name"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene1"</span>, <span class="st">"gene2"</span>, <span class="st">"gene3"</span>, <span class="st">"gene4"</span><span class="op">)</span>,</span>
<span>  <span class="st">"score"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.45</span>, <span class="op">-</span><span class="fl">3.21</span>, <span class="fl">2.44</span>, <span class="op">-</span><span class="fl">2.45</span><span class="op">)</span>,</span>
<span>  <span class="st">"group_label"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"group1"</span>, <span class="st">"group1"</span>, <span class="st">"group2"</span>, <span class="st">"group3"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="differential-expression-analysis-results-difexp">2.3. Differential expression analysis results (difexp)<a class="anchor" aria-label="anchor" href="#differential-expression-analysis-results-difexp"></a>
</h3>
<p>A differential expression dataframe is optional but
<strong>recommended</strong> if available. It facilitates downstream
signature extraction, and signature comparison by rank-based test.<br></p>
<p><code>difexp</code> is a dataframe with the following
<strong>required</strong> columns:<br>
“<strong>probe_id</strong>”, “<strong>feature_name</strong>”,
“<strong>score</strong>”, along with at least one of the following:
“<strong>p_value</strong>”, “<strong>q_value</strong>”, or
“<strong>adj_p</strong>”. Same as in the signature dataframe,
<strong>“group_label”</strong> is also required when the signature is
bi-directional or categorical.<br>
Descriptions of probe_id, feature_name, and group_label were provided in
the signature section above.<br>
“<strong>p_value</strong>”, “<strong>q_value</strong>”, or
“<strong>adj_p</strong>” refers to the p- or q-value representing the
significance of each feature.<br>
“<strong>score</strong>” is a numeric value that indicates the
importance or significance of a feature. Depending on how the signature
was derived, this can be the t-statistics, log-fold change, Z-score, or
other summary statistics.<br><br>
Here we use an example from the differential expression analysis using
the <code>limma</code> package.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Version reading from a txt file</span></span>
<span><span class="co"># difexp &lt;- read.table(</span></span>
<span><span class="co">#   file.path(</span></span>
<span><span class="co">#     system.file("extdata", package = "OmicSignature"),</span></span>
<span><span class="co">#     "difmatrix_Myc_mice_liver_24m.txt"</span></span>
<span><span class="co">#   ),</span></span>
<span><span class="co">#   header = TRUE, sep = "\t", stringsAsFactors = FALSE</span></span>
<span><span class="co"># )</span></span>
<span><span class="co"># Version reading from a binary file</span></span>
<span><span class="va">difexp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"OmicSignature"</span><span class="op">)</span>,</span>
<span>    <span class="st">"difmatrix_Myc_mice_liver_24m.rds"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">difexp</span><span class="op">)</span></span>
<span><span class="co">#&gt;   Probe.ID  logFC AveExpr      t P.Value adj.P.Val      b            ensembl</span></span>
<span><span class="co">#&gt; 1 10345228 -0.167   7.106 -1.470   0.186     0.560 -5.866 ENSMUSG00000103746</span></span>
<span><span class="co">#&gt; 2 10354534  0.041   4.351  0.520   0.620     0.870 -6.780 ENSMUSG00000060715</span></span>
<span><span class="co">#&gt; 3 10354529 -0.175   4.955 -0.941   0.379     0.731 -6.458 ENSMUSG00000043629</span></span>
<span><span class="co">#&gt; 4 10346337  0.025   8.621  0.188   0.857     0.962 -6.911 ENSMUSG00000038323</span></span>
<span><span class="co">#&gt; 5 10353792 -0.025   6.063 -0.284   0.785     0.936 -6.885 ENSMUSG00000045815</span></span>
<span><span class="co">#&gt; 6 10350848 -0.055   7.595 -0.630   0.549     0.836 -6.712 ENSMUSG00000049881</span></span>
<span><span class="co">#&gt;     gene_symbol</span></span>
<span><span class="co">#&gt; 1 1700001G17Rik</span></span>
<span><span class="co">#&gt; 2 1700019A02Rik</span></span>
<span><span class="co">#&gt; 3 1700019D03Rik</span></span>
<span><span class="co">#&gt; 4 1700066M21Rik</span></span>
<span><span class="co">#&gt; 5 1700101I19Rik</span></span>
<span><span class="co">#&gt; 6 2810025M15Rik</span></span></code></pre></div>
<p>Manually change the column names to match the requirement. The
built-in function <code><a href="../reference/replaceDifexpCol.html">replaceDifexpCol()</a></code> is designed to replace
<em>some</em> of the frequently used alternative column names.<br></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">difexp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/replaceDifexpCol.html">replaceDifexpCol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">difexp</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in replaceDifexpCol(colnames(difexp)): Required column for</span></span>
<span><span class="co">#&gt; OmicSignature object difexp: feature_name, is not found in your input. This may</span></span>
<span><span class="co">#&gt; cause problem when creating your OmicSignature object.</span></span>
<span><span class="va">difexp</span> <span class="op">&lt;-</span> <span class="va">difexp</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>feature_name <span class="op">=</span> <span class="va">ensembl</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>group_label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">score</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="st">"MYC Reduce"</span>, <span class="st">"WT"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">difexp</span><span class="op">)</span></span>
<span><span class="co">#&gt;   probe_id  logfc  mean  score p_value adj_p      b       feature_name</span></span>
<span><span class="co">#&gt; 1 10345228 -0.167 7.106 -1.470   0.186 0.560 -5.866 ENSMUSG00000103746</span></span>
<span><span class="co">#&gt; 2 10354534  0.041 4.351  0.520   0.620 0.870 -6.780 ENSMUSG00000060715</span></span>
<span><span class="co">#&gt; 3 10354529 -0.175 4.955 -0.941   0.379 0.731 -6.458 ENSMUSG00000043629</span></span>
<span><span class="co">#&gt; 4 10346337  0.025 8.621  0.188   0.857 0.962 -6.911 ENSMUSG00000038323</span></span>
<span><span class="co">#&gt; 5 10353792 -0.025 6.063 -0.284   0.785 0.936 -6.885 ENSMUSG00000045815</span></span>
<span><span class="co">#&gt; 6 10350848 -0.055 7.595 -0.630   0.549 0.836 -6.712 ENSMUSG00000049881</span></span>
<span><span class="co">#&gt;     gene_symbol group_label</span></span>
<span><span class="co">#&gt; 1 1700001G17Rik          WT</span></span>
<span><span class="co">#&gt; 2 1700019A02Rik  MYC Reduce</span></span>
<span><span class="co">#&gt; 3 1700019D03Rik          WT</span></span>
<span><span class="co">#&gt; 4 1700066M21Rik  MYC Reduce</span></span>
<span><span class="co">#&gt; 5 1700101I19Rik          WT</span></span>
<span><span class="co">#&gt; 6 2810025M15Rik          WT</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="create-the-omicsignature-object">2.4. Create the <code>OmicSignature</code> object<a class="anchor" aria-label="anchor" href="#create-the-omicsignature-object"></a>
</h3>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">OmS</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/OmicSignature.html">OmicSignature</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span></span>
<span>  metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>  signature <span class="op">=</span> <span class="va">signature</span>,</span>
<span>  difexp <span class="op">=</span> <span class="va">difexp</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt;   [Success] OmicSignature object Myc_reduce_mice_liver_24m created.</span></span></code></pre></div>
<p>Set <code>print_message</code> = <code>TRUE</code> to see the
messages.<br></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">OmS</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/OmicSignature.html">OmicSignature</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span></span>
<span>  metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>  signature <span class="op">=</span> <span class="va">signature</span>,</span>
<span>  difexp <span class="op">=</span> <span class="va">difexp</span>,</span>
<span>  print_message <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt;   --Required attributes for metadata: signature_name, phenotype, organism, direction_type, assay_type, author --</span></span>
<span><span class="co">#&gt;   [Success] Metadata is saved. </span></span>
<span><span class="co">#&gt;   [Success] Signature is valid. </span></span>
<span><span class="co">#&gt;   [Success] difexp is valid. </span></span>
<span><span class="co">#&gt;   [Success] OmicSignature object Myc_reduce_mice_liver_24m created.</span></span></code></pre></div>
<p>See the created object information:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">OmS</span><span class="op">)</span></span>
<span><span class="co">#&gt; Signature Object: </span></span>
<span><span class="co">#&gt;   Metadata: </span></span>
<span><span class="co">#&gt;     adj_p_cutoff = 0.05 </span></span>
<span><span class="co">#&gt;     assay_type = transcriptomics </span></span>
<span><span class="co">#&gt;     author = Researcher1 </span></span>
<span><span class="co">#&gt;     covariates = none </span></span>
<span><span class="co">#&gt;     description = mice Myc haploinsufficient (Myc(+/-)) </span></span>
<span><span class="co">#&gt;     direction_type = bi-directional </span></span>
<span><span class="co">#&gt;     keywords = Myc, KO, longevity </span></span>
<span><span class="co">#&gt;     organism = Mus musculus </span></span>
<span><span class="co">#&gt;     others = C57BL/6 </span></span>
<span><span class="co">#&gt;     phenotype = Myc_reduce </span></span>
<span><span class="co">#&gt;     platform = transcriptomics by array </span></span>
<span><span class="co">#&gt;     PMID = 25619689 </span></span>
<span><span class="co">#&gt;     sample_type = liver </span></span>
<span><span class="co">#&gt;     score_cutoff = 5 </span></span>
<span><span class="co">#&gt;     signature_name = Myc_reduce_mice_liver_24m </span></span>
<span><span class="co">#&gt;     year = 2015 </span></span>
<span><span class="co">#&gt;   Metadata user defined fields: </span></span>
<span><span class="co">#&gt;     animal_strain = C57BL/6 </span></span>
<span><span class="co">#&gt;   Signature: </span></span>
<span><span class="co">#&gt;     MYC Reduce (5)</span></span>
<span><span class="co">#&gt;     WT (10)</span></span>
<span><span class="co">#&gt;   Differential Expression Data: </span></span>
<span><span class="co">#&gt;     884 x 10</span></span></code></pre></div>
<p>Use new criteria to extract significant features:<br>
(this does <em>not</em> change the <code>signature</code> saved in the
object)</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">OmS</span><span class="op">$</span><span class="fu">extractSignature</span><span class="op">(</span><span class="st">"abs(score) &gt; 10; adj_p &lt; 0.01"</span><span class="op">)</span></span>
<span><span class="co">#&gt;   probe_id       feature_name   score group_label</span></span>
<span><span class="co">#&gt; 1 10349648 ENSMUSG00000004552  14.762  MYC Reduce</span></span>
<span><span class="co">#&gt; 2 10345762 ENSMUSG00000026072 -13.543          WT</span></span>
<span><span class="co">#&gt; 3 10353192 ENSMUSG00000025932  10.487  MYC Reduce</span></span>
<span><span class="co">#&gt; 4 10355259 ENSMUSG00000061816 -10.315          WT</span></span></code></pre></div>
<p>Besides saving and reading OmicSignature object in <code>.rds</code>
format, you can export the object as a text file in json format.<br></p>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">OmS</span>, <span class="st">"Myc_reduce_mice_liver_24m_OmS.rds"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/writeJson.html">writeJson</a></span><span class="op">(</span><span class="va">OmS</span>, <span class="st">"Myc_reduce_mice_liver_24m_OmS.json"</span><span class="op">)</span></span></code></pre>
<p>See more in <a href="https://montilab.github.io/OmicSignature/articles/FunOmS.html">“Functionalities
of OmicSignature”</a> section.<br></p>
</div>
</div>
<div class="section level2">
<h2 id="create-an-omicsignature-from-difexp-and-metadata">3. Create an <code>OmicSignature</code> from <code>difexp</code> and
<code>metadata</code><a class="anchor" aria-label="anchor" href="#create-an-omicsignature-from-difexp-and-metadata"></a>
</h2>
<p>You can by-pass the generating signature process once you are an
expert. Simply provide cutoffs (e.g. <code>adj_p_cutoff</code> and
<code>score_cutoff</code>) in the <code>metadata</code>, make sure
<code>difexp</code> has “adj_p”, “score” and “group_label” columns, and
use <code><a href="../reference/OmicSigFromDifexp.html">OmicSigFromDifexp()</a></code> to automatically extract
significant features and create the <code>OmicSignature</code>
object.<br></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">OmS1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/OmicSigFromDifexp.html">OmicSigFromDifexp</a></span><span class="op">(</span><span class="va">difexp</span>, <span class="va">metadata</span><span class="op">)</span></span>
<span><span class="co">#&gt; -- criterias used to extract signatures:  abs(score) &gt;= 5; adj_p &lt;= 0.05 . </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   [Success] OmicSignature object Myc_reduce_mice_liver_24m created.</span></span>
<span><span class="va">OmS1</span></span>
<span><span class="co">#&gt; Signature Object: </span></span>
<span><span class="co">#&gt;   Metadata: </span></span>
<span><span class="co">#&gt;     adj_p_cutoff = 0.05 </span></span>
<span><span class="co">#&gt;     assay_type = transcriptomics </span></span>
<span><span class="co">#&gt;     author = Researcher1 </span></span>
<span><span class="co">#&gt;     covariates = none </span></span>
<span><span class="co">#&gt;     description = mice Myc haploinsufficient (Myc(+/-)) </span></span>
<span><span class="co">#&gt;     direction_type = bi-directional </span></span>
<span><span class="co">#&gt;     keywords = Myc, KO, longevity </span></span>
<span><span class="co">#&gt;     organism = Mus musculus </span></span>
<span><span class="co">#&gt;     others = C57BL/6 </span></span>
<span><span class="co">#&gt;     phenotype = Myc_reduce </span></span>
<span><span class="co">#&gt;     platform = transcriptomics by array </span></span>
<span><span class="co">#&gt;     PMID = 25619689 </span></span>
<span><span class="co">#&gt;     sample_type = liver </span></span>
<span><span class="co">#&gt;     score_cutoff = 5 </span></span>
<span><span class="co">#&gt;     signature_name = Myc_reduce_mice_liver_24m </span></span>
<span><span class="co">#&gt;     year = 2015 </span></span>
<span><span class="co">#&gt;   Metadata user defined fields: </span></span>
<span><span class="co">#&gt;     animal_strain = C57BL/6 </span></span>
<span><span class="co">#&gt;   Signature: </span></span>
<span><span class="co">#&gt;     MYC Reduce (5)</span></span>
<span><span class="co">#&gt;     WT (10)</span></span>
<span><span class="co">#&gt;   Differential Expression Data: </span></span>
<span><span class="co">#&gt;     884 x 10</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="section"></h2>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Vanessa Mengze Li, Stefano Monti.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
